src/coreclr/jit/importercalls.cpp

```diff
@@ -1571,7 +1581,62 @@ var_types Compiler::impImportCall(OPCODE                  opcode,
 }
 
 //------------------------------------------------------------------------
-// impThrowIfNull: Remove redundandant boxing from ArgumentNullException_ThrowIfNull
+// impFoldEnumEquals: Optimize Enum.Equals calls by unboxing and comparing underlying values
+//
+// Arguments:
+//    call -- call representing Enum.Equals
+//
+// Return Value:
+//    Optimized tree (or nullptr if we can't optimize it).
+//
+GenTree* Compiler::impFoldEnumEquals(GenTreeCall* call)
+{
+    assert(call->IsSpecialIntrinsic(this, NI_System_Enum_Equals));
+    assert(call->AsCall()->gtArgs.CountUserArgs() == 2);
+
+    GenTree* arg0 = call->AsCall()->gtArgs.GetArgByIndex(0)->GetNode();
+    GenTree* arg1 = call->AsCall()->gtArgs.GetArgByIndex(1)->GetNode();
+
+    bool isArg0Exact;
+    bool isArg1Exact;
+    bool isNonNull; // we don't care about this property for both args here
+
+    CORINFO_CLASS_HANDLE cls0 = gtGetClassHandle(arg0, &isArg0Exact, &isNonNull);
+    CORINFO_CLASS_HANDLE cls1 = gtGetClassHandle(arg1, &isArg1Exact, &isNonNull);
+
+    if ((cls0 != cls1) || (cls0 == NO_CLASS_HANDLE) || !isArg0Exact || !isArg1Exact)
+    {
+        return nullptr;
+    }
+
+    CORINFO_CLASS_HANDLE underlyingEnumCls;
+    if (info.compCompHnd->isEnum(cls1, &underlyingEnumCls) != TypeCompareState::Must)
```

Comment:
Why is this check needed?

Reply:
> Why is this check needed?

Not really needed judging by the managed impl, although, it does have a `Debug.Assert(rt.IsActualEnum)` so I converted the check to an [assert](https://github.com/dotnet/runtime/blob/52746629166f04db4d8948f0244d002409edfa74/src/coreclr/System.Private.CoreLib/src/System/Enum.CoreCLR.cs#L20).